### PR TITLE
Add flag for Gas instrumentation.

### PR DESCRIPTION
**Pull request by kferles**
- Initial (and experimental) version of gas instrumentation.
- It extracts gas costs directly from solc. This feature in solc seems to be a bit limited, but it's better than nothing.
- For verisol integration, please use this form of solc: https://github.com/utopia-group/solidity

### CODE REVIEW

**Comment by shuvendu-lahiri** [comment]
It appears that the solc version has to be upgraded to use this feature from within VeriSol?

**Comment by kferles** [comment]
The only change I made to solc was to print gas costs for AST nodes when the output format is set to json. I can do the same for the current version that verisol works with. However, even if the json ast does not contain any gas cost information, this feature will still work but the instrumentation will be essentially a noop because it will initialize all gas costs to zero.

File: Sources/SolToBoogie/ProcedureTranslator.cs [hunk 1]
```diff
@@ -49,6 +78,7 @@ public ProcedureTranslator(TranslatorContext context, bool _genInlineAttrsInBpl
 
         public override bool Visit(ContractDefinition node)
         {
+            preTranslationAction(node);
```
Comment (shuvendu-lahiri) [inline Sources/SolToBoogie/ProcedureTranslator.cs]: Seems that you are invoking preTranslationAction for each node type. Can this not be folded into one statement in visitor?

Comment (kferles) [reply Sources/SolToBoogie/ProcedureTranslator.cs]: Yeah, that's true. I tried to come up with a more elegant solution to this but it does not seem to be possible in the current infrastructure. In general, performing default actions inside a visitor is known to be problematic.

Comment (kferles) [reply Sources/SolToBoogie/ProcedureTranslator.cs]: Another solution would be to perform the gas instrumentation in a separate pass. But in this case, I would need to keep a mapping from every solidity AST node to its first Boogie statement. Is there an easy way to calculate this?

**Pull request merged**